File: Daily/Count_Sub_Matricies_with_top_left_element_and_sum_less_than_k/Count_Sub_Matricies_with_top_left_element_and_sum_less_than_k.cpp
// Count_Sub_Matricies_with_top_left_element_and_sum_less_than_k.cpp
//
// Compile locally with:
// clang++ Count_Sub_Matricies_with_top_left_element_and_sum_less_than_k.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int countSubmatrices(vector<vector<int>> &grid, int k)
    {
        int height = grid.size();
        int width = grid[0].size();
        int count{0};
        vector<vector<int>> dp(height + 1, vector<int>(width + 1, 0));

        for (int y = 1; y < height + 1; y++)
        {
            for (int x = 1; x < width + 1; x++)
            {
                dp[y][x] = grid[y - 1][x - 1] + dp[y - 1][x] + dp[y][x - 1] - dp[y - 1][x - 1];
                if (dp[y][x] <= k)
                {
                    count++;
                }
            }
        }
        return count;
    }
};

int main()
{
    cout << "Running Count_Sub_Matricies_with_top_left_element_and_sum_less_than_k.cpp..." << endl;

    Solution sol;

    // TODO:
    // Add local test calls here

    return 0;
}

File: Daily/Count_Sub_Matricies_with_top_left_element_and_sum_less_than_k/count__sub__matricies_with_top_left_element_and_sum_less_than_k.py
from typing import Optional, List
import heapq
from collections import defaultdict


class Solution:
    def countSubmatrices(self, grid: List[List[int]], k: int) -> int:
        height = len(grid)
        width = len(grid[0])

        dp = [[0 for _ in range(width + 1)] for _ in range(height + 1)]

        count = 0
        for y in range(1, height + 1):
            for x in range(1, width + 1):
                dp[y][x] = (
                    grid[y - 1][x - 1] + dp[y - 1][x] + dp[y][x - 1] - dp[y - 1][x - 1]
                )
                if dp[y][x] <= k:
                    count += 1

        return count


if __name__ == "__main__":
    sol = Solution()
    grid = [[7, 2, 9], [1, 5, 0], [2, 6, 6]]
    k = 20
    print(sol.countSubmatrices(grid, k))

    grid = [[7, 6, 3], [6, 6, 1]]
    k = 18
    print(sol.countSubmatrices(grid, k))
    print("Running Solution...")
